design/baremetal-operator/user-defined-parameters.md

```diff
@@ -0,0 +1,456 @@
+                dirty = true
+        }
+
++       extraParametersChanged := !info.host.Status.TriedExtraParameters.Match(*info.bmcExtraParametersSecret)
++       if extraParametersChanged {
++               info.log.Info("new extra parameters")
++               info.host.UpdateTriedExtraParameters(*info.bmcExtraParametersSecret)
++               info.postSaveCallbacks = append(info.postSaveCallbacks, updatedExtraParameters.Inc)
++               dirty = true
++       }
++
+        preprovImgFormats, err := prov.PreprovisioningImageFormats()
+        if err != nil {
+                return actionError{err}
+@@ -796,7 +804,7 @@ func (r *BareMetalHostReconciler) registerHost(prov provisioner.Provisioner, inf
+                        PreprovisioningImage:  preprovImg,
+                        HasCustomDeploy:       hasCustomDeploy(info.host),
+                },
+-               credsChanged,
++               credsChanged || extraParametersChanged,
+                info.host.Status.ErrorType == metal3v1alpha1.RegistrationError)
+
+        if errors.Is(err, provisioner.ErrNeedsPreprovisioningImage) &&
+@@ -854,7 +862,7 @@ func (r *BareMetalHostReconciler) registerHost(prov provisioner.Provisioner, inf
+                }
+        }
+
+-       // Reaching this point means the credentials are valid and worked,
++       // Reaching this point means the credentials/extraParameters are valid and worked,
+        // so clear any previous error and record the success in the
+        // status block.
+        registeredNewCreds := !info.host.Status.GoodCredentials.Match(*info.bmcCredsSecret)
+@@ -863,11 +871,21 @@ func (r *BareMetalHostReconciler) registerHost(prov provisioner.Provisioner, inf
+                info.host.UpdateGoodCredentials(*info.bmcCredsSecret)
+                info.publishEvent("BMCAccessValidated", "Verified access to BMC")
+                dirty = true
+-       } else {
++       }
++
++       registeredNewExtraParameters := !info.host.Status.GoodExtraParameters.Match(*info.bmcExtraParametersSecret)
++       if registeredNewExtraParameters {
++               info.log.Info("updating extra parameters success status fields")
++               info.host.UpdateGoodExtraParameters(*info.bmcExtraParametersSecret)
++               info.publishEvent("BMCExtraParametersValidated", "Verified access to BMC")
++               dirty = true
++       }
++
++       if !registeredNewCreds && !registeredNewExtraParameters {
+                info.log.Info("verified access to the BMC")
+        }
+
+-       if info.host.Status.ErrorType == metal3v1alpha1.RegistrationError || registeredNewCreds {
++       if info.host.Status.ErrorType == metal3v1alpha1.RegistrationError || registeredNewCreds || registeredNewExtraParameters {
+                info.log.Info("clearing previous error message")
+                dirty = clearError(info.host)
+        }
+```
+
+### Risks and Mitigations
+
+None
```

Comment:
How will verification work? What if a user supplies an unsupported parameter? What if a user supplies an invalid value?

Reply:
We can't verify the parameters on the BMO side, so the verification work should be handled by ironic. We'll add that risk here.